juju/juju
# K8s resource storage fixes

## Description of change

There are 2 fixes here in 2 commits.
1. Add support for OCI docker images in the juju attach-resource command.
All of the backend APIs are already in place; the change is to the command itself. It makes an additional
API call to get the charm resource metadata and properly handles file resources vs OCI image resources.
There's a file rename which bloats the diff a bit.

2. Fix k8s storage provisioning - storage other than k8s (eg tmpfs, rootfs) was broken. Also, fix support for scaling applications to 0, as it had been disabled since storage relied on having on unit. Now when storage is provisioned, we don't rely on looking at unit storage metadata but get the storage info directly from the application metadata. This is much cleaner and less complicated.

## QA steps

bootstrap k8s
deploy mariadb-k8s
juju scale-application mariadb-k8s 0
(check k8s cluster that pods are gone, and also units gone)
juju scale-application mariadb-k8s 1
(check that a new unit and pod are created)
juju deploy mariadb-k8s --storage database=tmpfs
(kubectl describe pod to see that emptyDir storage is used)
juju attach-resource mariadb-k8s mysql_image=mariadb
(check mongo db that resource has been updated)
